File: src/lexer/Lexer.cpp
//
// Created by Lenovo on 2023/9/12.
//

#include <iostream>
#include "Lexer.h"
#include "../util/Exceptions.h"
#include <cctype>

std::vector<Token> Lexer::tokens = {};
std::string Lexer::content;

const std::unordered_map<std::string, std::string> Lexer::KEY_TYPE_MAP = {
        {"int", "INTTK"},
        {"printf", "PRINTFTK"},
        {"getint", "GETINTTK"},
        {"for", "FORTK"},
        {"const", "CONSTTK"},
        {"main", "MAINTK"},
        {"if", "IFTK"},
        {"else", "ELSETK"},
        {"void", "VOIDTK"},
        {"break", "BREAKTK"},
        {"continue", "CONTINUETK"},
        {"return", "RETURNTK"},
        {"+", "PLUS"},
        {"==", "EQL"},
        {"-", "MINU"},
        {"!=", "NEQ"},
        {"<", "LSS"},
        {"<=", "LEQ"},
        {">", "GRE"},
        {">=", "GEQ"},
        {"=", "ASSIGN"},
        {"/", "DIV"},
        {";", "SEMICN"},
        {"%", "MOD"},
        {",", "COMMA"},
        {"(", "LPARENT"},
        {")", "RPARENT"},
        {"[", "LBRACK"},
        {"]", "RBRACK"},
        {"{", "LBRACE"},
        {"}", "RBRACE"},
        {"||", "OR"},
        {"&&", "AND"},
        {"!", "NOT"},
        {"*", "MULT"},
};

static int position = 0;


Token Lexer::next() {
    std::string buf;
    while (isspace(content[position])) position++;
    for (; position < content.size();) {
        if (isAlpha(content[position])) {
            buf = getWord();
            return {KEY_TYPE_MAP.find(buf) != KEY_TYPE_MAP.end() ? std::string(KEY_TYPE_MAP.at(buf)) : "IDENFR", buf};
        } else if (isNum(content[position])) {
            buf = getNum();
            return {"INTCON", buf};
        } else {
            return getSym();
        }
    }
    return {"ERR", "ERR"};
    throw LexerException("Error at Lexer.next\n");
}

std::string Lexer::getWord() {
    std::string buf;
    while (position < content.size() && (isAlpha(content[position]) || isNum(content[position]) || content[position] == '_'))
        buf += content[position++];
    return buf;
}

std::string Lexer::getNum() {
    std::string buf;
    while (position < content.size() && isNum(content[position]))
        buf += content[position++];
    return buf;
}

Token Lexer::getSym() {
    if (content[position] == '/') {
        position++;
        if (content[position] == '/' || content[position] == '*') { // 注释
            skipComment();
            return next();
        } else {
            return {"DIV", "/"};
        }
    } else {
        position++;
        std::string buf;
        switch (content[position - 1]) {
            case '!':
            case '=':
            case '<':
            case '>':
                buf = content[position - 1];
                if (content[position] == '=') {
                    buf += content[position++];
                    return {std::string(KEY_TYPE_MAP.at(buf)), buf};
                } else {
                    return {std::string(KEY_TYPE_MAP.at(buf)), buf};
                }
            case '|':
            case '&':
                buf = std::string(2, content[position++]);
                return {std::string(KEY_TYPE_MAP.at(buf)), buf};
            case '"':
                return getStr();
            default:
                buf = content[position - 1];
                if (KEY_TYPE_MAP.find(buf) != KEY_TYPE_MAP.end())
                    return {std::string(KEY_TYPE_MAP.at(buf)), buf};
                else {
                    return {"ERR", "ERR"};
                    throw LexerException("Error at Lexer.getSym\n" + buf);
                }
        }
    }
}

void Lexer::skipComment() {
    if (content[position] == '/') {
        while (position < content.size() && content[position] != '\n')
            position++;
    } else {
        position++;
        while (position + 1 < content.size() && !(content[position] == '*' && content[position + 1] == '/'))
            position++;
        position += 2;
    }
}

Token Lexer::getStr() {
    std::string buf = "\"";
    while (content[position] != '"')
        buf += content[position++];
    buf += content[position++];
    return {"STRCON", buf};
}

void Lexer::run() {
    while (position < content.size()) {
        try {
            tokens.push_back(next());
        } catch(const std::out_of_range& e) {
            std::cout << e.what() << std::endl;
            std::cout << content[position - 1] << std::endl;
        }
    }
}

// Fast Inverse Square Root

File: src/lexer/Lexer.java
package lexer;

import util.GlobalConfigure;

import java.util.ArrayList;
import java.util.HashMap;

public class Lexer {
    public static final Lexer LEXER = new Lexer();
    private Lexer() {
        TYPE_MAP.put("main", TokenType.MAINTK);
        TYPE_MAP.put("const", TokenType.CONSTTK);
        TYPE_MAP.put("int", TokenType.INTTK);
        TYPE_MAP.put("break", TokenType.BREAKTK);
        TYPE_MAP.put("continue", TokenType.CONTINUETK);
        TYPE_MAP.put("if", TokenType.IFTK);
        TYPE_MAP.put("else", TokenType.ELSETK);
        TYPE_MAP.put("!", TokenType.NOT);
        TYPE_MAP.put("&&", TokenType.AND);
        TYPE_MAP.put("||", TokenType.OR);
        TYPE_MAP.put("for", TokenType.FORTK);
        TYPE_MAP.put("getint", TokenType.GETINTTK);
        TYPE_MAP.put("printf", TokenType.PRINTFTK);
        TYPE_MAP.put("return", TokenType.RETURNTK);
        TYPE_MAP.put("+", TokenType.PLUS);
        TYPE_MAP.put("-", TokenType.MINU);
        TYPE_MAP.put("void", TokenType.VOIDTK);
        TYPE_MAP.put("*", TokenType.MULT);
        TYPE_MAP.put("/", TokenType.DIV);
        TYPE_MAP.put("%",TokenType.MOD);
        TYPE_MAP.put("<", TokenType.LSS);
        TYPE_MAP.put("<=", TokenType.LEQ);
        TYPE_MAP.put(">", TokenType.GRE);
        TYPE_MAP.put(">=", TokenType.GEQ);
        TYPE_MAP.put("==", TokenType.EQL);
        TYPE_MAP.put("!=", TokenType.NEQ);
        TYPE_MAP.put("=", TokenType.ASSIGN);
        TYPE_MAP.put(";", TokenType.SEMICN);
        TYPE_MAP.put(",", TokenType.COMMA);
        TYPE_MAP.put("(", TokenType.LPARENT);
        TYPE_MAP.put(")", TokenType.RPARENT);
        TYPE_MAP.put("[", TokenType.LBRACK);
        TYPE_MAP.put("]", TokenType.RBRACK);
        TYPE_MAP.put("{", TokenType.LBRACE);
        TYPE_MAP.put("}", TokenType.RBRACE);
    }

    private final HashMap<String, TokenType> TYPE_MAP = new HashMap<>();

    int pos = 0;    // 当前处于字符串的位置
    int line = 1;
    int cnt = 0;    // 当前访问到哪个 token
    String content;
    ArrayList<Token> tokens = new ArrayList<>();

    public void run(String string) {
        content = string.trim();
        while (pos < content.length()) {
            addToken();
        }
    }

    public Token preView(int offset) {
        if (cnt + offset >= 0 && cnt + offset < tokens.size()) {
            return tokens.get(cnt + offset);
        }
        return null;
    }

    private void addToken() {
        while (pos < content.length() && isSpace(content.charAt(pos))) {
            if (content.charAt(pos) == '\n') {
                line++;
            }
            pos++;
        }
        if (pos > content.length()) {
            return;
        }
        if (isWordBegin(content.charAt(pos))) {
            tokens.add(getWord());
        } else if (Character.isDigit(content.charAt(pos))) {
            tokens.add(getDigit());
        } else {
            Token token = getSym();
            if (token != null) {
                tokens.add(token);
            }
        }
    }

    private Token getWord() {
        StringBuilder builder = new StringBuilder();
        while (isWord(content.charAt(pos))) {
            builder.append(content.charAt(pos));
            pos++;
        }
        return new Token(TYPE_MAP.getOrDefault(builder.toString(), TokenType.IDENFR),
                builder.toString(), line);
    }

    private Token getDigit() {
        StringBuilder builder = new StringBuilder();
        while (Character.isDigit(content.charAt(pos))) {
            builder.append(content.charAt(pos));
            pos++;
        }
        return new Token(TokenType.INTCON, builder.toString(), Integer.parseInt(builder.toString()), line);
    }

    private Token getSym() {
        StringBuilder builder = new StringBuilder();
        builder.append(content.charAt(pos));
        if (content.charAt(pos) == '/') {
            pos++;
            if (content.charAt(pos) == '/') {
                pos++;
                skipComment(1);
            } else if (content.charAt(pos) == '*') {
                pos++;
                skipComment(2);
            } else {
                return new Token(TokenType.DIV, builder.toString(), line);
            }
            return null;
        } else {
            if ("!=<>".indexOf(content.charAt(pos)) != -1) {
                pos++;
                if (content.charAt(pos) == '=') {
                    builder.append(content.charAt(pos));
                    pos++;
                }
                return new Token(TYPE_MAP.get(builder.toString()), builder.toString(), line);
            } else if ("|&".indexOf(content.charAt(pos)) != -1) {
                pos++;
                builder.append(content.charAt(pos));
                pos++;
                return new Token(TYPE_MAP.get(builder.toString()), builder.toString(), line);
            } else if (content.charAt(pos) == '"') {
                return getStr();
            } else {
                pos++;
                return new Token(TYPE_MAP.get(builder.toString()), builder.toString(), line);
            }
        }
    }

    private void skipComment(int type) {    // 1: //   2: /**/
        if (type == 1) {
            while (pos < content.length() && content.charAt(pos) != '\n') {
                pos++;
            }
        } else {
            while (pos + 1 < content.length() && !(content.charAt(pos) == '*' && content.charAt(pos + 1) == '/')) {
                if (content.charAt(pos) == '\n') {
                    line++;
                }
                pos++;
            }
            pos += 2;
        }
    }

    private Token getStr() {
        StringBuilder builder = new StringBuilder("\"");
        pos++;
        while (pos < content.length() && content.charAt(pos) != '"') {
            builder.append(content.charAt(pos));
            pos++;
        }
        builder.append(content.charAt(pos++));
        if (GlobalConfigure.ERROR && isLegal(builder.toString())) {

        }
        return new Token(TokenType.STRCON, builder.toString(), line);
    }

    public boolean isLegal(String buf) {
        for (int i = 1; i < buf.length() - 1; i++) {
            char ch = buf.charAt(i);
            if ((ch == '\\' && buf.charAt(i + 1) != 'n') || (ch == '%' && buf.charAt(i + 1) != 'd')) {
                return false;
            }
            if (!(ch == '%' || ch == 32 || ch == 33 || ch >= 40 && ch <= 126)) {
                return false;
            }
        }
        return true;
    }

    private boolean isWordBegin(char ch) {
        return Character.isLetter(ch) || ch == '_';
    }

    private boolean isWord(char ch) {
        return isWordBegin(ch) || Character.isDigit(ch);
    }

    public String printThis() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Token token: tokens) {
            stringBuilder.append(token.toString()).append('\n');
        }
        return stringBuilder.toString();
    }

    private boolean isSpace(char ch) {
        return ch == '\r' || ch == '\n' || ch == ' ' || ch == '\t';
    }

}
